generics/middleware/authenticator.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * name : middleware/authenticator.js
+ * author : Aman Karki
+ * Date : 20-July-2020
+ * Description : Keycloak authentication.
+ */
+
+// dependencies
+const jwtDecode = require('jwt-decode');
+const slackClient = require("../helpers/slack-communications");
+const interceptor = require("./lib/api-interceptor");
+const messageUtil = require("./lib/message-util");
+let responseCode = require("../http-status-codes");
+const shikshalokam = require("../helpers/shikshalokam");
+
+const apiInterceptor = require('./apiInterceptor');
+
+var reqMsg = messageUtil.REQUEST;
+
+
+var respUtil = function (resp) {
+  return {
+    status: resp.errCode,
+    message: resp.errMsg,
+    currentDate: new Date().toISOString()
+  };
+};
+
+var tokenAuthenticationFailureMessageToSlack = function (req, token, msg) {
+  let jwtInfomration = jwtDecode(token)
+  jwtInfomration["x-authenticated-user-token"] = token
+  const tokenByPassAllowedLog = { 
+    method: req.method, 
+    url: req.url, 
+    headers: req.headers, 
+    body: req.body,
+    errorMsg: msg, 
+    customFields: 
+    jwtInfomration, 
+    slackErrorName: process.env.SLACK_ERROR_NAME,
+    color: process.env.SLACK_ERROR_MESSAGE_COLOR
+  }
+
+  slackClient.sendMessageToSlack(tokenByPassAllowedLog)
+}
+
+var removedHeaders = [
+  "host",
+  "origin",
+  "accept",
+  "referer",
+  "content-length",
+  "accept-encoding",
+  "accept-language",
+  "accept-charset",
+  "cookie",
+  "dnt",
+  "postman-token",
+  "cache-control",
+  "connection"
+];
+
+async function getAllRoles(obj) {
+  let roles = await obj.roles;
+  await _.forEach(obj.organisations, async value => {
+    roles = await roles.concat(value.roles);
+  });
+  return roles;
+}
+
+
+
+module.exports = async function (req, res, next,token="") {
+
+  removedHeaders.forEach(function (e) {
+    delete req.headers[e];
+  });
+
+  var token = req.headers["x-authenticated-user-token"];
+  if (!req.rspObj) req.rspObj = {};
+  var rspObj = req.rspObj;
+
+  if (req.path.includes("slack")) {
+    next();
+    return
+  }
+
+
+  // Allow search endpoints for non-logged in users.
+  let guestAccessPaths = ["bodh/search","bodh/request"];
+  await Promise.all(guestAccessPaths.map(async function (path) {
+    if (req.path.includes(path)) {
+      next();
+      return;
+    }
+  }));
+
+  let internalAccessApiPaths = ["token/verify","keywords"];
+  let performInternalAccessTokenCheck = false;
```

Comment:
@rakeshSgr - Handle case of internal access token and x-authenticated user token both are required.